scripts/istat/geos/preprocess.py

```diff
@@ -0,0 +1,96 @@
+    "Cadastral code of the municipality", "Legal population 2011 (09/10/2011)", "NUTS1", "NUTS2(3)", "NUTS3"]
+  data.to_csv(FILE_PATH_en, encoding = 'utf-8', index = False)
+  
+def find_duplicates(df):
+  #find not one-to-one mapping from ISTAT to NUTS
+  duplicated = pd.DataFrame(columns = df.columns)
+  for col in df.columns:
+    dup_vals = df[df[col].duplicated()][col].unique()
+    for val in dup_vals:
+      duplicated = pd.concat([duplicated, df[df[col]==val]], ignore_index = True)
+  duplicated = duplicated.drop_duplicates()
+  print(duplicated)
+  return duplicated
+      
+def preprocess():
+
+  data = pd.read_csv(FILE_PATH_en)
+  columns_rename = {"Province Code (Historic) (1)": "Province Code","Name of the supra-municipal territorial unit (valid for statistical purposes)":"Province name", "Automotive abbreviation":"Province Abbreviation", "NUTS2(3)":"NUTS2","Common Code numeric format":"Municipal Code", "Name in Italian":"Municipal Name"}
+  data = data.rename(columns = columns_rename)
+  
+  #correct some of the mismatch of NUTS code and names
+  reorg = [("ITG2A", 91, "OG", "Ogliastra"), ("ITG28", 95, "OR", "Oristano"), ("ITG27", 92, "CA","Cargliari"),
+       ("ITG29", 90, "OT", "Olbia-Tempio")]
+  for (nuts3, province_code, province_abbrev, province_name) in reorg:
+    data.loc[data[data["NUTS3"] == nuts3].index, "Province Code"] = province_code
+    data.loc[data[data["NUTS3"] == nuts3].index, "Province Abbreviation"] = province_abbrev
+    data.loc[data[data["NUTS3"] == nuts3].index, "Province name"] = province_name
+  data.loc[data[data["Province name"] == "Napoli"].index, "Province Abbreviation"] = "NA"
+  
+  region_data = data[["Region Code", "NUTS2", "Region name"]].drop_duplicates()
+  region_data["NUTS2"] = "nuts/" + region_data["NUTS2"]
+  region_data["Region Code"] = region_data["Region Code"].astype(str).str.zfill(2)
+  region_data.loc[region_data[region_data["NUTS2"] == "nuts/ITH1"].index, "Region name"] = "Provincia Autonoma di Bolzano/Bozen"
+  region_data.loc[region_data[region_data["NUTS2"] == "nuts/ITH2"].index, "Region name"] = "Provincia Autonoma di Trento"
+  region_data.to_csv("ISTAT_region.csv", index = False)
+  
+  province_data = data[["Province Code", "NUTS3","Province name", "Province Abbreviation"]].drop_duplicates()
+  province_data["NUTS3"] = "nuts/" + province_data["NUTS3"]
+  province_data["Province Code"] = province_data["Province Code"].astype(str).str.zfill(3)
+  province_data.to_csv("ISTAT_province.csv", index = False)
+  
+  municipal_data = data[["Municipal Code", "Municipal Name", "NUTS3"]].drop_duplicates()
+  municipal_data["NUTS3"] = "dcid:nuts/" + municipal_data["NUTS3"]
+  municipal_data["Municipal Code"] = municipal_data["Municipal Code"].astype(str).str.zfill(6)
+  municipal_data.to_csv("ISTAT_municipal.csv", index = False)
+
+def generate_tmcf():
+  region_TEMPLATE = "Node: E:ISTAT->E0\n" +\
```

Comment:
We don't have to do autogeneration of template MCF. If it's just 3, it's perfectly find to just handwrite them. We can remove this part.

Reply:
done